src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs

```diff
@@ -348,8 +348,11 @@ internal NamespaceOrTypeOrAliasSymbolWithAnnotations BindNamespaceOrTypeOrAliasS
                             {
                                 // Inside a method body or other executable code, we can pull on NonNullTypes symbol or question IsValueType without causing cycles.
                                 // Types created outside executable context should be checked by the responsible symbol (the method symbol checks its return type, for instance).
-
-                                if (!typeArgument.IsValueType && NonNullTypesContext.NonNullTypes != true)
+                                if (Flags.Includes(BinderFlags.SuppressNullableContraintChecks))
+                                {
+                                    diagnostics.Add(new LazyMissingNonNullTypesContextDiagnosticInfo(NonNullTypesContext, typeArgument), nullableSyntax.QuestionToken.GetLocation());
+                                }
+                                else if (!typeArgument.IsValueType && NonNullTypesContext.NonNullTypes != true)
                                 {
                                     diagnostics.Add(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, nullableSyntax.QuestionToken.GetLocation());
```

Comment:
>WRN_MissingNonNullTypesContextForAnnotation [](start = 62, length = 43)

We should try and factor this check into `ReportNullableReferenceTypesIfNeeded` as well. I'll handle in my next PR.